File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter14/Petzold.Windows8.VectorDrawing/Petzold.Windows8.VectorDrawing/LineSegment.cs
using System.Collections.Generic;
using Windows.Foundation;

namespace Petzold.Windows8.VectorDrawing
{
    public struct LineSegment : IGeometrySegment
    {
        readonly Point point1, point2;
        readonly double a, b;         // as in x = ay + b
        
        public LineSegment(Point point1, Point point2) : this()
        {
            this.point1 = point1;
            this.point2 = point2;

            a = (point2.X - point1.X) / (point2.Y - point1.Y);
            b = point1.X - a * point1.Y;
        }

        public void GetAllX(double y, IList<double> xCollection)
        {
            if ((point2.Y > point1.Y && y >= point1.Y && y < point2.Y) ||
                (point2.Y < point1.Y && y <= point1.Y && y > point2.Y))
            {
                xCollection.Add(a * y + b);
            }
        }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter14/PetzoldWindows8VectorDrawing/PetzoldWindows8VectorDrawing/LineSegment.cpp
#include "pch.h"
#include "LineSegment.h"

using namespace PetzoldWindows8VectorDrawing;

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;

void LineSegment::Set(Point point1, Point point2)
{
    this->point1 = point1;
    this->point2 = point2;

    a = (point2.X - point1.X) / (point2.Y - point1.Y);
    b = point1.X - a * point1.Y;
}

void LineSegment::GetAllX(float y, IVector<float>^ xCollection)
{
    if ((point2.Y > point1.Y && y >= point1.Y && y < point2.Y) ||
        (point2.Y < point1.Y && y <= point1.Y && y > point2.Y))
    {
        xCollection->Append(a * y + b);
    }
}
